[MTPG] Use TLS propagation to enable MTPG from bwd.

kumpera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104735

We use PyTorch's built-in tls propagation in ThreadLocalState to forward the world object
from the fwd thread to the bwd thread.

This further closes the gap on enabling FSDP.

kumpera committed:
[MTPG] Use TLS propagation to enable MTPG from bwd.

We use PyTorch's built-in tls propagation in ThreadLocalState to forward the world object
from the fwd thread to the bwd thread.

This further closes the gap on enabling FSDP.

[ghstack-poisoned]

rohan-varma commented on test/distributed/test_multi_threaded_pg.py:
```diff
@@ -249,7 +252,25 @@ def test_all_reduce_coalesced(self):
+    @skip_if_lt_x_gpu(1)
+    def test_bwd_sees_fwd_pg(self):
+
+        class MyFunc(torch.autograd.Function):
```
can we have a test where we explicitly retrieve the process groups used in fwd and bwd and validate they are equivalent?

wconstab replied:
actually, what is "forward_pg"? The test name implies there is a pg used in forward, but the forwrad function doesn't do any distributed compute so is `forward_pg` the same as `default pg`?

kumpera replied:
The global PG of the fwd thread should match the one of the bwd thread.

Should be pretty easy to test for that, I'll change the test to do so once I figure out why it's crashing.

rohan-varma reviewed (approved):
LGTM, will test this on FSDP!

wconstab commented on torch/testing/_internal/distributed/multi_threaded_pg.py:
```diff
@@ -389,13 +389,19 @@ class WorldData:
 
     def _get_world(self) -> WorldData:
+        if torch._C._is_key_in_tls(MTPG_TLS_KEY):
+            return torch._C._get_obj_in_tls(MTPG_TLS_KEY)
```
are we trying to ensure 2 threads (e.g. forward and autograd) receive the same object for a `_get_world()` call?  If so i am confused about how the code helps, it looks like we are allowing a TLS version of _get_world() which would allow divergence

kumpera replied:
I'm not 100% sure because there's some bad memory management going on, I'm trying to use the same pattern as the cuda_graph code but it's not working at all for me.

kumpera replied:
This is no longer relevant as I changed the design.

awgu commented:
Test error looks real :/

https://github.com/pytorch/pytorch/actions/runs/5480357846/jobs/9983604919
```
distributed/test_multi_threaded_pg 1/1 failed! Received signal: SIGSEGV
Traceback (most recent call last):
  File "test/run_test.py", line 1683, in <module>
    main()
  File "test/run_test.py", line 1652, in main
    failure_messages = run_tests(
  File "test/run_test.py", line 1580, in run_tests
    raise RuntimeError(err_message)
RuntimeError: distributed/test_multi_threaded_pg 1/1 failed! Received signal: SIGSEGV
```

kumpera committed:
Update on "[MTPG] Use TLS propagation to enable MTPG from bwd."


We use PyTorch's built-in tls propagation in ThreadLocalState to forward the world object
from the fwd thread to the bwd thread.

This further closes the gap on enabling FSDP.

[ghstack-poisoned]

kumpera commented:
@pytorchmergebot merge

kumpera commented:
@pytorchmergebot merge -f "cudnn failure is unrelated."
